File: Server/src/main/java/server/Service.java
package server;

import model.domain.Agency;
import model.domain.Reservation;
import model.domain.Trip;
import model.utils.MyException;
import services.IObserver;
import services.IServices;
import services.ServiceException;

import java.sql.Time;
import java.util.Collection;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class Service implements IServices {
    private final AgencyService agencySrv;
    private final TripService tripSrv;
    private final ReservationService resSrv;
    private final Map<String, IObserver> loggedAgencies;
    private final int defaultThreadsNo = 5;

    public Service(AgencyService agencySrv, TripService tripSrv, ReservationService resSrv) {
        this.agencySrv = agencySrv;
        this.tripSrv = tripSrv;
        this.resSrv = resSrv;
        loggedAgencies = new ConcurrentHashMap<>();
    }

    @Override
    public synchronized void saveReservation(Reservation reservation) throws ServiceException {
        try {
            resSrv.saveReservation(reservation);
        } catch (MyException e) {
            throw new ServiceException("Client already has a reservation to this trip");
        }
        ExecutorService executor = Executors.newFixedThreadPool(defaultThreadsNo);

        loggedAgencies.forEach((a, o) -> {
            if (!a.equals(reservation.getAgency()))
                executor.execute(() -> {
                    try {
                        System.out.println("Notifying [" + a + "] - new reservation saved");
                        o.ReservationSaved(reservation);
                    } catch (ServiceException e) {
                        System.out.println("Error notifying agency " + a);
                    }
                });
        });
    }

    @Override
    public synchronized void login(Agency agency, IObserver observer) throws ServiceException {
        Agency agency2 = agencySrv.getAgency(agency);
        if (agency2 != null) {
            if (loggedAgencies.get(agency.getName()) != null)
                throw new ServiceException("Agency already logged in");
            loggedAgencies.put(agency.getName(), observer);
        } else
            throw new ServiceException("Authentication failed");
    }

    @Override
    public synchronized void logout(Agency agency, IObserver observer) throws ServiceException{
        IObserver localObserver = loggedAgencies.remove(agency.getName());
        if (localObserver == null)
            throw new ServiceException("Agency " + agency.getName() + "is not logged in");
    }

    @Override
    public synchronized Collection<Agency> getAgencies() {
        return agencySrv.getAllAgencies();
    }

    @Override
    public synchronized Collection<Trip> getTrips(String destination, Time startTime, Time endTime) {
        return tripSrv.getTouristAttractionTrips(destination, startTime, endTime);
    }

    @Override
    public synchronized Collection<Reservation> getReservations() { return resSrv.getAllReservations(); }

}

File: Server/server/Service.cs
using System;
using System.Collections.Generic;
using System.Threading.Tasks;
using Model;
using Services;

namespace Server.server
{
    public class Service : IServices
    {
        private readonly AgencyService _agencySrv;
        private readonly TripService _tripSrv;
        private readonly ReservationService _resSrv;
        private readonly IDictionary<string, IObserver> _loggedAgencies;

        public Service(AgencyService agencySrv, TripService tripSrv, ReservationService resSrv)
        {
            _agencySrv = agencySrv;
            _tripSrv = tripSrv;
            _resSrv = resSrv;
            _loggedAgencies = new Dictionary<string, IObserver>();
        }

        public void Login(Agency agency, IObserver observer)
        {
            Agency agency2 = _agencySrv.Get(agency);
            if (agency2 != null)
            {
                if (_loggedAgencies.ContainsKey(agency.Name))
                    throw new ServiceException("Agency already logged in");
                _loggedAgencies[agency.Name] = observer;
            }
            else
                throw new ServiceException("Authentication failed");
        }

        public void Logout(Agency agency, IObserver observer)
        {
            IObserver localObserver = _loggedAgencies[agency.Name];
            if (localObserver == null)
                throw new ServiceException("Agency " + agency.Name + "is not logged in");
            _loggedAgencies.Remove(agency.Name);
        }

        public IEnumerable<Agency> GetAgencies() => _agencySrv.GetAll();

        public IEnumerable<Trip> GetTrips(string destination, TimeSpan startTime, TimeSpan endTime) =>
            _tripSrv.GetTouristAttractionTrips(destination, startTime, endTime);

        public IEnumerable<Reservation> GetReservations() => _resSrv.GetAllReservations();

        public void SaveReservation(Reservation reservation)
        {
            try
            {
                _resSrv.SaveReservation(reservation);
            }
            catch (ArgumentException e)
            {
                throw new ServiceException(e.Message);
            }

            foreach (KeyValuePair<string, IObserver> entry in _loggedAgencies)
                if (entry.Key != reservation.Agency)
                    Task.Run(() => entry.Value.ReservationSaved(reservation));
        }
    }
}
